### PR TITLE
update seed_worker to set seed based on worker_id and rank

### LINKED ISSUE #37932: Different DataLoader worker share the same seed and lost randomness
### System Info

train_dataloader use `seed_worker` to set the seed on different workers. But when working with distributed training, e.g. DDP, deep speed, etc, different ranks will shared the same seed. If there are random data augmentation, they will be the same across different ranks.

https://github.com/huggingface/transformers/blob/2932f318a20d9e54cc7aea052e040164d85de7d6/src/transformers/trainer.py#L1031

### Who can help?

@zach-huggingface @SunMarc

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

print(os.environ["LOCAL_RANK"], torch.randint(low=0, high=len(self), size=()).item())

in the `__getitem__ ` function of dataset, and find same random value across different rank

### Expected behavior

expect different value across different ranks

### PR DESCRIPTION
**Pull request by gathierry**
# What does this PR do?

Fixes #37932

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @sunmarc

**Review by SunMarc — APPROVED** [review]
Thanks ! Can you add a test ?

**Comment by gathierry** [comment]
Test added

**Comment by gathierry** [comment]
Hi @SunMarc, there are still 2 workflows pending, is it expected? Any additional actions needed on my side?
thanks

File: tests/trainer/test_trainer_distributed_worker_seed.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        execute_subprocess_async(cmd, env=self.get_env())
+
+
+def run_distributed_training(training_args):
+    set_seed(42)
+    model = DummyModel()
+    dataset = DummyDataset()
+    training_args.max_steps = 10
+    # dataloader_num_workers must be > 0 to enable worker_init_fn
+    training_args.dataloader_num_workers = 2
```
Comment (SunMarc) [inline tests/trainer/test_trainer_distributed_worker_seed.py:75]: minor nit, please set the output_dir to `self.get_auto_remove_tmp_dir()` so that we don't store the results

**Review by SunMarc — APPROVED** [review]
Thanks, looks good, just a final nit

File: tests/trainer/test_trainer_distributed_worker_seed.py (hunk 1 above)
Comment (gathierry) [reply tests/trainer/test_trainer_distributed_worker_seed.py:77]: Done. I put it in `test_trainer` as `args`

Comment (SunMarc) [reply tests/trainer/test_trainer_distributed_worker_seed.py:77]: thx ! merging
